src/storage/sql_table.cpp

```diff
@@ -1,34 +1,324 @@
+  }
+
+  // The versions are different
+  // 1. Check if we can just update the old version
+  // 2. If Yes:
+  //    2.a) Convert ProjectedRow into old ProjectedRow
+  //    2.b) Update the old DataTable using the old ProjectedRow
+  // 3. Else:
+  //    3.a) Get the old row
+  //    3.b) Convert it into new row
+  //    3.c) Delete old row
+  //    3.d) Update the new row before insert
+  //    3.e) Insert new row into new table
+
+  // Check if the Redo's attributes are a subset of old schema so that we can update old version in place
+  bool is_subset = true;
+
+  std::vector<catalog::col_oid_t> redo_col_oids;  // the set of col oids the redo touches
+  for (auto &it : map) {
+    redo_col_oids.emplace_back(it.first);
+    // check if the col_oid exists in the old schema
+    if (tables_.Find(old_version)->second.column_map.count(it.first) == 0) {
+      is_subset = false;
+      break;
+    }
+  }
+
+  storage::TupleSlot ret_slot;
+  if (is_subset) {
+    // we can update in place
+
+    // We should create a buffer of old Projected Row and update in place. We can't just
+    // directly erase the data without creating a redo and update the chain.
+
+    auto old_pair = InitializerForProjectedRow(redo_col_oids, version_num);
+
+    // 1. Create a ProjectedRow Buffer for the old version
+    byte *buffer = common::AllocationUtil::AllocateAligned(old_pair.first.ProjectedRowSize());
+    storage::ProjectedRow *pr = old_pair.first.InitializeRow(buffer);
+
+    // 2. Copy from new ProjectedRow to old ProjectedRow
+    StorageUtil::CopyProjectionIntoProjection(redo, map, tables_.Find(version_num)->second.layout, pr, old_pair.second);
+
+    // 3. Update the old data-table
+    bool result = tables_.Find(old_version)->second.data_table->Update(txn, slot, *pr);
+    delete[] buffer;
+    if (!result) {
+      return {false, slot};
+    }
+    ret_slot = slot;
+  } else {
+    STORAGE_LOG_DEBUG("have to delete and insert ... ");
+
+    // need to create a new ProjectedRow of all columns
+    // 1. Get the old row
+    // 2. Convert it into new row
+    // 3. Delete old row
+    // 4. Update the new row before insert
+    // 5. Insert new row into new table
+
+    // 1. Get old row
+    // 2. Convert it into new row
+    std::vector<catalog::col_oid_t> new_col_oids;  // the set of col oids which the new schema has
+    for (auto &it : tables_.Find(version_num)->second.column_map) new_col_oids.emplace_back(it.first);
+    auto new_pair = InitializerForProjectedRow(new_col_oids, version_num);
+    auto new_buffer = common::AllocationUtil::AllocateAligned(new_pair.first.ProjectedRowSize());
+    ProjectedRow *new_pr = new_pair.first.InitializeRow(new_buffer);
+    bool valid = Select(txn, slot, new_pr, new_pair.second, version_num);
+    if (!valid) {
+      delete[] new_buffer;
+      return {false, slot};
+    }
+    // 3. Delete the old row
+    bool succ = tables_.Find(old_version)->second.data_table->Delete(txn, slot);
+
```

Comment:
Consider an interleaving: here another thread begins a scan and runs to finish. The row being updated will be missing in the scan result. I haven't seen any mechanism that prevents this interleaving from happening.

Reply:
MVCC prevents this interleaving.  Either the delete and insert are uncommitted when the scan starts and it reads the version that is to be "deleted", or the scan started after the commit and sees the updated version.

Reply:
Is it possible that the scan and the update belongs to the same transactions so that MVCC won't help?

Reply:
This behavior would be undefined in the current version since only installing the undo record is atomic.  Any concurrent read/update under the same transaction’s context would have the same problem in the non-versioned schema setting.